### ISSUE TITLE
Complex object validation

### ISSUE THREAD

**Issue by romaintha** [encode/apistar#272]
HI all, 
As far as I see, using the current typesystem, it is not possible to implement more complex object validation. Given this :
`
class MyObject(typesystem.Object):

    properties = {
        'a': typesystem.string(),
        'b': typesystem.string(),
        'c': typesystem.string(),
    }
`
I would like to have the following validation : anyOf ("a", allOf("b", "c")) (either "a" or "b" and "c") 
I could of course implement the validation logic myself, but then this would not be introspect for the documentation. So here is my question, do I miss something? Or is it something planned?

**Comment by tomchristie** [comment]
Yes we do need to provide hooks for more complex validations (also see database-dependant validations). I'm not sure how much of that would be declarative validations that can be mapped to schema information, and how much would just be custom server-side validation, but there's certainly some more work to be done in this area.

**Comment by romaintha** [comment]
So here is another naive question: why not using JSON schema for describing expected inputs/outputs?

**Comment by tomchristie** [comment]
We kinda *are* using JSON schema, in that the type system maps onto it. It's just that we don't yet support it's full feature set (and might not choose to do so.)

**Comment by imanhodjaev** [comment]
@tomchristie when do you estimate to add validation hook?

Thanks.

**Comment by imanhodjaev** [comment]
One thing to extend might `typesystem.Object` https://github.com/encode/apistar/blob/master/apistar/typesystem.py#L146 but I am unsure, feels like a bad idea to directly extend it.

**Comment by imanhodjaev** [comment]
I don't know if it is a good approach but defining validation component and custom validation method in schema seems natural.

In my case I have the following schema:

```python
# schemas.py
class StartEndTimestamp(Object):
    properties: dict = {
        'start_timestamp': integer(default=0),
        'end_timestamp': integer(default=0)
        # ...
        # ...
        # ...
    })
```
And I wanted to validate if `start < end` and return some error if not.
So I did the following things:

1. Define `error_list` field in `Object` schema
2. Define is valid method which make additional validation I want,
3. Use inside my request handler

```python
# schemas.py
class StartEndTimestamp(ts.Object):
    error_list: List[str]

    properties: dict = {
        'start_timestamp': integer(default=0),
        'end_timestamp': integer(default=0)
        # ...
        # ...
        # ...
    })

    def is_valid(self):
        if self['start_timestamp'] < self['end_timestamp']:
            self.error_list.append("Start timestamp must always be less than end timestamp")

        return len(self.error_list) == 0
```

```
# views.py

def filter_by_timestamp(filters: StartEndTimestamp):
    if filters.is_valid():
        # Return some stuff
    else:
        # Return 
```

At the moment I've been looking into raising exception if not valid and providing `validators` list field inside the schema and aggregating validation errors.

This approach I think is not so clean because schema by its design holds data and adding additional business logic seems not good.

@romaintha ping.
@tomchristie I need your feedback on this approach.

Thanks.
